[Concept]: Add bools

meatball133 opened the pull request:
resolves #343 

This pr is made to add bools to the crystal syllabus tree.

@angelikatyborska
@glennj

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+You are in process of developing the new highly appreciated game **Crystal Hunter**.
+In the game you are a character which moves around and collects crystals, when the player has picked up all crystals they win.
+Although if the player come in contact with a bandit so will the player lose all crystals and lose.
+There is an expectation to this rule and that is when the player has an active power up, which will make the player invisible for the bandits.
```
```suggestion
There is an exception to this rule: the player can have an active power-up which makes them invisible to the bandits. 
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+
+## 2. Define if character scores
+
+In the game will the character gets points when taking up a crystal or picking up a power up.
```
```suggestion
In the game, the player gets points when picking up a crystal or a power up.
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+
+## 1. Define if character gets bonus points
+
+In the game will the character get bonus points if they touch a bandit while having a power up.
```
```suggestion
In the game, the character will get bonus points if they touch a bandit while having a power-up.
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+
+In the game will the character get bonus points if they touch a bandit while having a power up.
+
+Define the `bonus_points?` method that takes two arguments (_if the character has an active power up_ and _if the character is touching a bandit_) and returns a boolean value if the Character will get bonus points. The method should return true only if the character has a power up active and is touching a bandit.
```
```suggestion
Define the `bonus_points?` method that takes two arguments (_if the character has an active power up_ and _if the character is touching a bandit_) and returns a boolean value that tells whether the character will get bonus points or not. The method should return `true` only if the character has a power up active and is touching a bandit, and `false` otherwise.
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+
+In the game will the character gets points when taking up a crystal or picking up a power up.
+
+Define the `score?` method that takes two arguments (_if the character is touching a power up_ and _if the character is touching a crystal_) and returns a boolean value if the character scored. The method should return true if the character is touching a power up or a crystal.
```
```suggestion
Define the `score?` method that takes two arguments (_if the character is touching a power up_ and _if the character is touching a crystal_) and returns a boolean value indicating if the character scored or not. The method should return `true` if the character is touching a power-up or a crystal, and return `false` otherwise.
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+
+## 3. Define if character loses
+
+Define the `lose?` method that takes two arguments (_if the character has a power up active_ and _if the character is touching a bandit_) and returns a boolean value if the character loses. The method should return true the character is touching a bandit and does not have a power up active.
```
```suggestion
Define the `lose?` method that takes two arguments (_if the character has a power up active_ and _if the character is touching a bandit_) and returns a boolean value that indicates if the character loses or not. The method should return `true` if the character is touching a bandit and does not have a power up active, and return `false` otherwise.
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,48 @@
+
+## 4. Define if character wins
+
+Define the `win?` method that takes three arguments (_if the character has eaten all of the crystals_, _if the character has a power up active_, and _if the character is touching a bandit_) and returns a boolean value if the character wins. The method should return true if the character has gathered all crystals and has not lost based on the arguments defined in part 3.
```
```suggestion
Define the `win?` method that takes three arguments (_if the character has eaten all of the crystals_, _if the character has a power up active_, and _if the character is touching a bandit_) and returns a boolean value indicating if the character wins or not. The method should return `true` if the character has gathered all crystals and has not lost based on the arguments defined in part 3, and return `false` otherwise.
```

andrerfcsantos commented on exercises/concept-wip/crystal-hunter/.meta/config.json:
```diff
@@ -0,0 +1,12 @@
+  },
+  "forked_from": ["elixir/pacman-rules"],
+  "icon": "low-power-embedded-game",
+  "blurb": "Learn about booleans by implementing the rules of the game, Crystal Hunter."
```
```suggestion
  "blurb": "Learn about booleans by implementing the rules of the Crystal Hunter game."
```

I think the comma is not needed here.

andrerfcsantos reviewed (commented):
This is exciting!

Made some suggestions for the things I think could be changed, let me know what you think. Full disclaimer: I know nothing about Crystal, but I contributed and wrote some concepts myself before.

For the prose, I focused more on reviewing the `about.md`, but some comments there might apply to the other files too.

Mostly reviewed the markdown files and ignored the Crystal files altogether -  it's best for someone more familiar with crystal than me to look at those.

angelikatyborska commented on concepts-wip/bools/.meta/config.json:
```diff
@@ -0,0 +1,5 @@
+{
+  "blurb": "Crystal has a Bool type to represent true and false and a bool can only be true or false. The bool type support AND, OR, and NOT operators",
```
In many languages, whether a type's name is uppercase or lowercase has a semantic difference. I would suggest to choose a consistent casing and be pedantic about always using that casing. Even if it doesn't matter in Crystal, it will make it easier on learners.

I think the official documentation uses uppercase for types, so:
```suggestion
  "blurb": "Crystal has a Bool type to represent true and false and a Bool can only be true or false. The Bool type support AND, OR, and NOT operators",
```

(This feedback probably applies to more than just this file)

meatball133 replied:
Yes, you are right. All types in crystal is either a module, class or struct (and sometimes a combination). The naming scheme for those are PascalCased.

angelikatyborska commented on exercises/concept-wip/crystal-hunter/.meta/src/exemplar.cr:
```diff
@@ -0,0 +1,17 @@
+class Rules
+  def eat_ghost?(power_pellet_active, touching_ghost)
+    power_pellet_active && touching_ghost
+  end
```
All the tests and example solution needs to be rewritten to actually use the new non-pacman terminology that you used in the instructions. After you do that, I will be able to follow the instructions and check if they make sense by solving the exercise myself.

angelikatyborska commented on exercises/concept-wip/crystal-hunter/.docs/introduction.md:
````diff
@@ -0,0 +1,116 @@
+Crystal has a type know as [`Bool`][bools], it is used to represent the values `true` and `false`.
+With the bool type can you use logical operators(`!`, `||`, `&&`) which can be used to combine bools and change their values.
+
+```crystal
+true && true
+# => True
+
+true && false
+# => False
+
+false && false
+# => false
+
+false || true
+# => True
+false || false
+# => false
+
+!false
+# => true
+
+!true
+# => false
+```
````
One of the feedback that @iHiD gave me about writing concept exercise introductions is that they should limit code examples to minimum. The goal is to force students to learn the text and really understand it without giving them the option to mindlessly copy-paste everything.

meatball133 replied:
Well I am not sure if I agree with Jeremy on that one. I think forcing the student to learn by text could make learning difficult since some students will be visual learners rather than textual. And have easier-to-follow code rather than reading text. With that said I think having text also is valuable. If there is a location where I could be more detailed, so let me know. For the earlier concepts I have written (not for crystal), I got praised by other maintainers on exercism for including a lot of examples. Thereby I have a bit mixed feelings about this.

Also another problem is that crystals documentation is not that good and is something the devs are aware of. If it is covered in the "doc" section, then is the doc good/okay but if it isn't it is hard to find good references.
I have gotten quite used to reading the source code of the language, but I do not think a lot of developers read the source code to understand how to do stuff.

I get the point, of reducing examples, they take place and some students like very consisted content so it is hard to make a concept that suits everybody.

It is again something I will have to think over, the snippet you mentioned has been reduced.

iHiD replied:
I think the point Angelika is making here is that by providing all the pairs written out, there's a risk the exercise becomes utterly trivial (basically people can just copy/paste) things. And then someone doesn't actually learn anything. It's important that someone can struggle a little to solve the exercises in order to actually learn.

> since some students will be visual learners rather than textual

Just on this note, there's pretty strong consensus among experts that different learning styles are a myth, which is something I was interested to learn and has changed my approach to creating content (e.g. https://fee.org/articles/learning-styles-don-t-actually-exist-studies-show/). 

I think probably **everyone** learns better with visuals and I am totally pro having code and visuals that demonstrate something. However, we have to be careful not just to give answers away and to create an appropriate amount of struggle to provide break-thought mental model moments.

---

Note: I'm not commenting on this specific example (I've not read the PR) but on the general principle.

meatball133 replied:
I think it is a valuable point both of you are bringing. I agree that the concepts should not be written in a way to allow "copy-pasting". Although as I tried to point out if you can't actually figure out what to do, could lead to frustrations. As I pointed out earlier crystal doesn't have that deeply written documentation when approaching difficult concepts. But if the student doesn't fully learn a concept, so will that likely lead to frustrations later. It is a bit of a balance to make the student think enough for themself, but not too much to make it frustrating. Thereby is it important to reveal just the right amount of information or as I would say in Swedish "lagom".

If there is a part where you think I am too "revling" or give too little information for that side too, please let me know and I will look over it/ you can suggest an improvement.

That with learning styles could be discussable, the article you sent is a debate article, and I agree that everyone could probably learn equally with every method. That doesn't mean that people don't have a preference for different learning styles. I dislike learning programming through youtube videos, although there are programming tutorial videos with 40 million views so somebody has to watch them, I rather have interactive learning. Meanwhile, when I learned chemistry I loved learning through videos and hated laborations (I was afraid to get some dangerous substance on my body). Although that doesn't mean that other students didn't appreciate those.

angelikatyborska commented on exercises/concept-wip/crystal-hunter/.docs/introduction.md:
````diff
@@ -0,0 +1,116 @@
+# => true
+
+true && false
+# => false
+
+false && false
+# => false
+```
+
+### Or(`||`)
+
+The [or operator][or] in crystal is represented by `||` and returns `true` if **at least one** of values given is `true` if both of the values are `false` then it returns `false`.
+When using the or operator so should one bool be placed one the right side of the `||` and one on the left side.
+
+```crystal
+true || true
+# => true
+
+true || false
+# => true
+
+false || true
+# => true
+
+false || false
+# => false
+```
+
+### Not(`!`)
+
+The not operator in crystal is represented by `!` and returns `true` if the given bool is `false` and return `false` if `true` is given.
+When using the not operator so should the bool be placed after the operator (`!`).
+
+```crystal
+!true
+# => false
+
+!false
+# => true
+```
````
All of this is a more detailed repetition of the content right after the first h1 heading ("Bool"). I would avoid repetitions and try to keep the introduction as short as possible (while still giving the students all the knowledge they need to solve the exercise) because people's patience towards reading is very short on the internet 🙂

meatball133 replied:
I decided to keep the first part but shorten the example there to make the first part shorter. I hope that improves it.

angelikatyborska replied:
Honestly, I still think the documents (about/introductions) are too long and too detailed. The way you have written them, they are great for somebody that has never heard about logical operators and is learning programming for the first time. But that is explicitly not the target group of Exercism. Exercism is targeting people who can already code in one language and want to learn another. For this reason, concept introductions should mention details that are unique to the programing language, but don't need to explain basic programming concepts that appear in every programming language.

meatball133 replied:
I mean it is a valid point, you know if you are a ruby developer, I would tell you to skip this concept because this doesn't teach anything new. The only difference between ruby and crystal is: Crystal doesn't have `and` and `or`. Ruby sees `false` and `nil` as false. I think Crystal sees `false`, `nil`, and `Nop` as false. I am a bit unsure about `Nop`, since it has to do with the compiler and is nothing you actually deal with in normal programming. These differences are kinda minor.

I guess a problem for me is that "Exercism is targeting people who can already code in one language", say a language does something totally different and the way Crystal may do it is "normal" for most languages but not in the perspective of that language. And sure I know around 8 languages, but all of them are quite similar, sure they all have some unique parts, but the most "out-pointing" languages are Elixir and Javascript, Javascript because weakly typed, and Elixir because functionally oriented. It feels hard to give a good perspective if you get what I mean.

I guess the whole explaining part under parentheses could be dropped and under `or` and `and` go to 2 examples instead of 3/4. Do you think such a change is satisfactory?

angelikatyborska commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,144 @@
+## Using brackets(`()`)
+
+When working with booleans you can use brackets to decided which bools to evaluate first.
+Dependent on how the brackets are used so can the result end up to be different.
```
Simpler grammatical structure to say the same thing:
```suggestion
The result can differ depending on how the brackets are used.
```

angelikatyborska commented on concepts-wip/bools/about.md:
````diff
@@ -0,0 +1,144 @@
+# => true
+```
+
+## Using brackets(`()`)
````
You might want to use the word "parentheses" everywhere (singular: parenthesis, plural: parentheses). Technically parentheses are a type of brackets, but parentheses is a more specific term that most people use to describe those characters `()`

meatball133 commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,144 @@
+
+## Truthy and falsy
+
+In crystal so are there [truthy and falsy values][truthy_and_falsey_values].
```
I read a discussion online about falsy and falsey and that it was such a new word so the spelling was decided or something. I think the meaning is the same. But I agree that we should follow the same spelling as the official doc

meatball133 commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,144 @@
+
+In Crystal is it common to name method that returns a bool with a question mark (`?`).
+It is to show that the method will either return a truthy or falsy value.
+That is done to show that the method can be used with for example an if statement.
```
Hmmmm, the most common one I use is `includes?`, but I think explaining that method doesn't make sense.

I think I think it maybe make sense to replace that `zero?` method with `nil?`, since that is a real method in crystal.

meatball133 replied:
```crystal
def exchange(value)
    if value.nil?
        value = 5
    end
    value
end

exchange(10)
# => 10

exchange(nil)
# => 5
```

Do you think something like this make sense?

andrerfcsantos replied:
Sounds good!

meatball133 commented on concepts-wip/bools/introduction.md:
```diff
@@ -0,0 +1,28 @@
+# Bools
```
On the track I started to learn the "fundamentals" of exercism, so do they follow the same scheme as I do in this concept. I am at the same time aware of other tracks doing the same method as you mention. I double-checked the exercism doc, and my interpretation is that it is written in a bit of a way that both methods are acceptable if that makes sense.

I will have to think about this one a bit.

meatball133 replied:
I have thought about it for a bit, I have decided to follow your suggestion on this one. 

I will bring the change to make it up to shape later today.

meatball133 commented on concepts-wip/bools/about.md:
````diff
@@ -0,0 +1,144 @@
+
+```exercism/caution
+You should only use brackets when they affect the result, otherwise should they be omitted.
+```
````
I have a bit of mixed feelings about this, I have seen on exercism quite mixed use of both ways of writing. And the markdown spec says you can use backticks.

I think following the official way is great, although the use as I said is quite mixed across the track.
And some stuff I write for crystal is being very lightly reviewed (for example the docs for the generator, some stuff is being more highly reviewed and thereby would not apply to those) then it makes sense to have a formatter that can catch for example double spaces. The formatter I use does not like tildas to be used that way. And removes them.

I will have to think over this, I think tildas could be a good way of doing it, since it makes the "blocs" different from code examples in the markdown file. But I want my formatter so yeah, let me think about it.

andrerfcsantos replied:
If backticks work, I think it's fine to use them. I was just wondering if they would work at all.

meatball133 replied:
No I think it is great you pointed this out, and yes they do work for just text.

But when wanting a code block inside of them, am I a bit unsure, 4 backticks maybe make it work. But I am pretty sure 3 won't. So in any case I will have to do a change to make it consistent across the track.

meatball133 commented:
I have updated the files based on the feedback, I think I included all of your feedback, if it isn't in I probably commented on it. If I didn't and thereby missed it let me know.

I added both of you as contributors, double check so I spelled the names correctly since I am unsure if I did.

meatball133 commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,144 @@
+Dependent on how the brackets are used so can the result end up to be different.
+
+In crystal so is what is in brackets evaluated first.
+After brackets has been evaluated so are booleans evaluated from left to right.
```
I try to say that what is in the parentheses is evaluated first, then crystal takes the first (from the left) logical operator and evaluates it.

Does that make sense, is there any way I could improve the explanation there?

meatball133 commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,144 @@
+## Method naming conventions with booleans
+
+In Crystal is it common to name method that returns a bool with a question mark (`?`).
+It is to show that the method will either return a truthy or falsy value.
```
I think I improved this, plz let me know if I didn't

meatball133 commented:
Thanks for your feedback!✨

I added you as a contributor, as with everybody else, double check so I spelled your name correctly.

angelikatyborska commented:
I just wanted to let you know that I'm not ignoring the PR and I'll take another look at it on Saturday or Sunday, I'm too busy until then!

meatball133 commented:
No worries, as I said there is no hurry,

My plan is to merge in basic before this concept any way and that pr is still marked as a draft.

ryanplusplus commented on concepts-wip/bools/about.md:
````diff
@@ -0,0 +1,136 @@
+When working with booleans you can use parentheses to decided which Bools to evaluate first.
+The result can differ depending on how the parentheses are used.
+
+In crystal, what is in parentheses is evaluated first.
+After parentheses has been evaluated so are booleans evaluated from left to right.
+In the example below the _and_ operation be evaluated first, which will evaluate to: `false && false || true`.
+Then the second _and_ operation will be evaluated, which makes it look like: `false || true`, which in turn results in `true`.
+In the second example, it will first evaluate what is in the parenthesis. After this evaluation, the expression look like: `true && false && true`.
+In a situation where there is only one _and_ operation, all Bools have to be `true` to make the result be `true`.
+If one of the arguments to _and_ is `false` the result will also be `false`.
+
+```crystal
+true && false && false || true
+# => true
+
+true && false && (false || true)
+# => false
+```
````
```suggestion
In crystal, what is in parentheses is evaluated first.
After parentheses has been evaluated so are booleans evaluated from left to right.

In the example below the _and_ operation be evaluated first, which will evaluate to: `false && false || true`.
Then the second _and_ operation will be evaluated, which makes it look like: `false || true`, which in turn results in `true`.

``crystal
true && false && false || true
# => true
``

In the next example, it will first evaluate what is in the parenthesis. After this evaluation, the expression look like: `true && false && true`.

``crystal
true && false && (false || true)
# => false
``

In a situation where there is only one _and_ operation, all Bools have to be `true` to make the result be `true`.
If one of the arguments to _and_ is `false` the result will also be `false`.
```

I think it might make sense to interleave the examples and the discussion. I deliberately broke the code fences for the examples because they don't play nicely with the suggestion syntax, but hopefully you get the intent.

ryanplusplus commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
````diff
@@ -0,0 +1,53 @@
+The method should return `true` if the character has gathered all crystals and has not lost based on the arguments defined in part 3, and return `false` otherwise.
+
+```crystal
+win?(false, true, false)
````
```suggestion
Rules.win?(false, true, false)
```

ryanplusplus commented on exercises/concept-wip/crystal-hunter/.docs/introduction.md:
```diff
@@ -0,0 +1,95 @@
+
+## Logical operators
+
+Crystal has 3 logical operators which is used to combine Bools and make expressions that produce different values.
```
```suggestion
Crystal has 3 logical operators which are used to combine Bools and make expressions that produce different values.
```

ryanplusplus commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,136 @@
+
+## Logical operators
+
+Crystal has 3 logical operators which is used to combine Bools and make expressions that produce different values.
```
```suggestion
Crystal has 3 logical operators which are used to combine Bools and make expressions that produce different values.
```

ryanplusplus commented on concepts-wip/bools/introduction.md:
```diff
@@ -0,0 +1,95 @@
+
+## Logical operators
+
+Crystal has 3 logical operators which is used to combine Bools and make expressions that produce different values.
```
```suggestion
Crystal has 3 logical operators which are used to combine Bools and make expressions that produce different values.
```

ryanplusplus commented on exercises/concept-wip/crystal-hunter/spec/crystal_hunter_spec.cr:
```diff
@@ -0,0 +1,64 @@
+
+describe Rules do
+  describe "bonus_points?" do
+    it "Chracter gets bonus points" do
+      Rules.new.bonus_points?(true, true).should be_true
+    end
+
+    it "Chracter gets no bonus points because no power-up is active" do
+      Rules.new.bonus_points?(false, true).should be_false
+    end
+
+    it "Chracter gets no bonus points because not touching bandit" do
+      Rules.new.bonus_points?(true, false).should be_false
+    end
+
+    it "Chracter gets no bonus points because no power-up is active and character is not touching bandit" do
```
```suggestion
    it "Character gets bonus points" do
      Rules.new.bonus_points?(true, true).should be_true
    end

    it "Character gets no bonus points because no power-up is active" do
      Rules.new.bonus_points?(false, true).should be_false
    end

    it "Character gets no bonus points because not touching bandit" do
      Rules.new.bonus_points?(true, false).should be_false
    end

    it "Character gets no bonus points because no power-up is active and character is not touching bandit" do
```

ryanplusplus commented on exercises/concept-wip/crystal-hunter/.meta/src/exemplar.cr:
```diff
@@ -0,0 +1,17 @@
+class Rules
+  def bonus_points?(power_pellet_active, touching_bandit)
```
Is there a reason that these are instance methods instead of class methods?

meatball133 replied:
It makes so we don't have to explain what self does. As fast as we reach the class concept, I think that should be explained, and that concept appears quite early.

angelikatyborska commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
```diff
@@ -0,0 +1,53 @@
+# Instructions
```
The instructions are clear and correct, and following them, I was able to easily solve the exercise without looking up any hints and without cheating 👍

angelikatyborska commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,136 @@
+When working with booleans you can use parentheses to decided which Bools to evaluate first.
+The result can differ depending on how the parentheses are used.
+
+In crystal, what is in parentheses is evaluated first.
+After parentheses has been evaluated so are booleans evaluated from left to right.
```
I have noticed that you tend to do a lot of small grammar mistakes. Have you considered using some sort of automated grammar checker to help you out?

I sometimes use Grammarly. It's not perfect, and it doesn't understand how programmers talk about programming... but it's better than nothing. Here's an example of how Grammarly could help you improve those two sentences:

<img width="1309" alt="Screenshot 2023-02-25 at 13 15 28" src="https://user-images.githubusercontent.com/7852553/221356574-5cde45d8-5f66-49cb-9745-c9c8908bccaf.png">

It doesn't find all the problems, but it's a good start. Here's how I would fix the grammar even further (but I'm not a linguist or a native speaker either)

```suggestion
In Crystal, what is in parentheses is evaluated first.
After parentheses have been evaluated, the booleans are evaluated from left to right.
```

meatball133 replied:
I used Grammarly for my earlier concepts and for writing this message XD. I have a spell checker connected to my ide. But I will use Grammarly from now on(So let me write another commit with those changes), but if you see any grammar mistakes(After the commit), please let me know.

glennj commented:
@meatball133  My same comment as the Numbers concept: any diffs between about.md/instructions.md/instructions.md ?

meatball133 commented:
both instructions.md/instructions.md is the same, about.md includes more info such as truthy and falsey and why methods ends with a `?`

glennj commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,126 @@
+# Bools
+
+Crystal has a type known as [`Bool`][bools], it is used to represent the values `true` and `false`.
+You can use logical operators (`!`, `||`, `&&`) which can be used to combine Bools and create more complex expressions:
```
```suggestion
```
You're just about to introduce these. I don't think the repetition is needed. I suggest deleting.
(same in introduction)

glennj commented on concepts-wip/bools/about.md:
````diff
@@ -0,0 +1,126 @@
+# => false
+```
+
+In a situation where there is only one _and_ operation, all Bools have to be `true` to make the result be `true`.
+If one of the arguments to _and_ is `false` the result will also be `false`.
````
You already said this in the `&&` section.

glennj commented on concepts-wip/bools/about.md:
````diff
@@ -0,0 +1,126 @@
+# => true
+```
+
+```exercism/caution
+You should only use parentheses when they affect the result, otherwise, should they be omitted.
+```
````
This is a style note, not a caution; use `exercism/note`

https://exercism.org/docs/building/markdown/markdown

glennj commented on concepts-wip/bools/about.md:
```diff
@@ -0,0 +1,126 @@
+true && nil
+# => false
+
+true && 4
```
```suggestion
false || 4
```

glennj commented on exercises/concept-wip/crystal-hunter/.docs/hints.md:
```diff
@@ -0,0 +1,25 @@
+
+## General
+
+- Don't worry about how the arguments are _derived_, focus on combining the arguments to return the intended result.
```
What does this mean? Are the values derived in some confusing way?

meatball133 replied:
Tbh I just copied that from the exercise I borrowed it from, not 100% sure personally, but I think it is about that you don't have to think about the logic if the character is touching the crystal since that logic is already given to you.

glennj commented on exercises/concept-wip/crystal-hunter/.meta/design.md:
```diff
@@ -0,0 +1,27 @@
+
+The goal of this exercise is to teach the student about Bools in crystal.
+
+## Learning objectives
+
+- Know what a Bool is (True and False)
+- Know about truthy and falsy
+- Know how to use `!`, `&&`, and `||`
+- Know that you can affect the result of Bool with brackets.
+- Know about the naming scheme of having `?` at the end of a method name
+
+## Out of scope
+
+- Using comparison operators
+- If statements
```
Objectives are met, within scope. Well done!

glennj commented on exercises/concept-wip/crystal-hunter/.docs/instructions.md:
````diff
@@ -0,0 +1,53 @@
+The method should return `true` if the character is touching a power-up or a crystal, and return `false` otherwise.
+
+```crystal
+Rules.score?(true, true)
````
Are there power-up crystals? Shouldn't be both true I think

meatball133 replied:
There are power-ups and crystal nonrelated, you get points for picking up either.

glennj commented on exercises/concept-wip/crystal-hunter/spec/crystal_hunter_spec.cr:
```diff
@@ -0,0 +1,64 @@
+    it "Don't lose if not touching a bandit" do
+      Rules.new.lose?(false, false).should be_false
+    end
+  end
```
Might as well add the "with power-up not touching bandit" case
```suggestion
    it "Don't lose if powered-up and not touching a bandit" do
      Rules.new.lose?(true, false).should be_false
    end
  end
```

glennj reviewed (changes requested):
I like it. Just a few suggestions.

meatball133 commented:
I feel this pr has got plenty of feedback so if nobody has any further feedback. I planned to merge this pr soon,

ErikSchierboom reviewed (approved):
Really nice! My only suggestion would be to remove the `-wip` suffix from the directories.
